### PR TITLE
Patch hdd images before migration

### PR DESCRIPTION
**Pull request by dgutu-suse**
Based on - https://progress.opensuse.org/issues/10996
Couple of examples:
http://corr.arch.suse.de/tests/1048
http://corr.arch.suse.de/tests/944
http://wingenfelder.arch.suse.de/tests/44
This PR will also require changes on testcases related to offline migrations.

### CODE REVIEW

File: lib/bootloader_setup.pm [hunk 1]
```diff
@@ -65,6 +65,7 @@ sub select_bootmenu_option {
     }
     if (get_var('ZDUP') || get_var("ONLINE_MIGRATION")) {
         send_key 'ret';    # boot from hard disk
+        assert_screen 'grub2', 15;
```
Comment (okurz) [inline lib/bootloader_setup.pm]: why?

File: lib/registration.pm [hunk 2]
```diff
@@ -201,6 +201,7 @@ sub fill_in_registration_data {
             assert_screen("module-selection");
             send_key $cmd{next};
         }
+        else {send_key $cmd{next};}
```
Comment (okurz) [inline lib/registration.pm]: that does not make any sense. If the same key is pressed in both "if" as well as "else", it should be outside the if-else.
OTOH why? ;-)

File: lib/utils.pm [hunk 3]
```diff
@@ -213,7 +213,7 @@ sub fully_patch_system {
 sub minimal_patch_system {
     my (%args) = @_;
     $args{version_variable} //= 'VERSION';
-    if (sle_version_at_least('12-SP1', version_variable => $args{version_variable})) {
+    if (sle_version_at_least('12-SP1', version_variable => $args{version_variable}) && !check_var('HDDVERSION', 'SLES-11')) {
```
Comment (okurz) [inline lib/utils.pm]: I don't think this is a good idea. It already tests for sle_version_at_least with a forward of the type of variable so it should do exactly what you want to do already

File: lib/utils.pm [hunk 4]
```diff
@@ -331,19 +331,26 @@ sub ensure_shim_import {
...
         assert_screen 'reboot-auth';
         wait_still_screen 3;
         type_password undef, max_interval => 5;
         # Extra assert_and_click (with right click) to check the correct number of characters is typed and open up the 'show text' option
+    if (!check_var('HDDVERSION', 'SLES-11')) {
         assert_and_click 'reboot-auth-typed', 'right';
         assert_and_click 'reboot-auth-showtext';         # Click the 'Show Text' Option to enable the display of the typed text
         assert_screen 'reboot-auth-correct-password';    # Check the password is correct
-
+    }
```
Comment (okurz) [inline lib/utils.pm]: that is messed up style

File: products/sle/main.pm [hunk 5]
```diff
@@ -969,7 +969,12 @@ sub load_fips_tests_crypt() {
     loadtest "console/cryptsetup";
     loadtest "console/ecryptfs_fips";
 }
-
+sub load_patching_tests() {
+    loadtest 'boot/boot_to_desktop_sym';
+    loadtest "update/patch_before_migration";
+    loadtest "console/consoletest_finish_sym";
+    loadtest "x11/reboot_and_install";
+}
```
Comment (okurz) [inline products/sle/main.pm]: please use only 'single ticks' with loadtest unless you want to use variable expansion

File: tests/autoyast/autoyast_reboot.pm [hunk 6]
```diff
@@ -25,8 +25,12 @@ sub run {
...
 
     #obsoletes installation/autoyast_reboot.pm
-    assert_screen("bios-boot",     900);
-    assert_screen("autoyast-boot", 20);
+    assert_screen("bios-boot", 900);
+
+    if (check_var("BOOTFROM", "d")) {
+        assert_screen("inst-bootmenu", 60);
+    }
+    else { assert_screen("autoyast-boot", 20); }
```
Comment (okurz) [inline tests/autoyast/autoyast_reboot.pm]: the change in this file seems unrelated. Please move into another commit.

File: tests/update/patch_before_migration.pm [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+# Maintainer: Dumitru Gutu <dgutu@suse.de>
+
+use base "consoletest";
+use strict;
+use testapi;
+use utils;
+use registration;
+
+sub patching_sle11() {
+
```
Comment (okurz) [inline tests/update/patch_before_migration.pm]: delete blank line

File: tests/update/patch_before_migration.pm [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    else {
+        $regopts = "-a regcode-sles=$regcode";
+    }
+    $registration = script_output("suse_register -n -a email=$email $regopts", 90);
+
+    die "Unable to register the system, please check logs" unless $registration =~ /Registration finished successfully/;
+    save_screenshot;
+    assert_script_run('zypper lr -d');
+
+    #Patch the system
```
Comment (okurz) [inline tests/update/patch_before_migration.pm]: redundant comment, the next function call has a good name already :-)

File: tests/update/patch_before_migration.pm [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    minimal_patch_system();
+    save_screenshot;
+    my $reg_out = script_output("suse_register -E");
+    die "Unable to erase system registration data" unless $reg_out =~ /Successfully erased local registration data/;
+    save_screenshot;
+    script_run("gconftool-2 --set /apps/gnome-packagekit/frequency_get_updates --type=string daily");
```
Comment (okurz) [inline tests/update/patch_before_migration.pm]: why?

File: tests/update/patch_before_migration.pm [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    assert_script_run('zypper lr -d');
+}
+
+sub patching_sle12() {
+
```
Comment (okurz) [inline tests/update/patch_before_migration.pm]: delete blank line

File: tests/update/patch_before_migration.pm [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+    minimal_patch_system();
+    save_screenshot;
+    my $reg_out = script_output("suse_register -E");
+    die "Unable to erase system registration data" unless $reg_out =~ /Successfully erased local registration data/;
+    save_screenshot;
+    script_run("gconftool-2 --set /apps/gnome-packagekit/frequency_get_updates --type=string daily");
+    assert_script_run('zypper lr -d');
+}
+
+sub patching_sle12() {
```
Comment (okurz) [inline tests/update/patch_before_migration.pm]: isn't that like a generic patch function, i.e. `patching_sle`?

File: tests/update/patch_before_migration.pm [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+        assert_script_run('rm /etc/zypp/credentials.d/* /etc/SUSEConnect');
+        my $output = script_output 'SUSEConnect -s';
+        die "System is still registered" unless $output =~ /Not Registered/;
+        save_screenshot;
+    }
+    assert_script_run("zypper mr --enable --all");
+}
+
+sub run() {
+
```
Comment (okurz) [inline tests/update/patch_before_migration.pm]: delete blank line

File: tests/x11/reboot_and_install.pm [hunk 13]
```diff
@@ -15,14 +15,16 @@ use strict;
...
 use testapi;
 use utils 'reboot_gnome';
-use bootloader_setup qw(select_bootmenu_option bootmenu_default_params);
+use bootloader_setup;
+use registration;
 
 sub run() {
     reboot_gnome;    # reboot from previously booted hdd to do pre check or change e.g. before upgrade
-    select_bootmenu_option;
+    return if select_bootmenu_option == 3;
```
Comment (okurz) [inline tests/x11/reboot_and_install.pm]: what does '3' mean?

File: tests/x11/reboot_and_install.pm [hunk 14]
```diff
@@ -15,14 +15,16 @@ use strict;
...
     bootmenu_default_params;
-    # boot
-    send_key "ret";
+    specific_bootmenu_params;
+    registration_bootloader_params(utils::VERY_SLOW_TYPING_SPEED);
```
Comment (okurz) [inline tests/x11/reboot_and_install.pm]: why?

File: tests/x11/reboot_and_install.pm [hunk 15]
```diff
@@ -15,14 +15,16 @@ use strict;
...
+    send_key "ret";    #boot
```
Comment (okurz) [inline tests/x11/reboot_and_install.pm]: better keep it atop the current line as before

**Comment by dgutu-suse** [comment]
Thanks for comment, I'll took a look what needs to be changed.
Yes, the changes are not small, and potentially something can fail but since I've modified files which has effect on other testcases like yast and online migrations, I have checked their run.

File: lib/bootloader_setup.pm (hunk 1 above)
Comment (dgutu-suse) [inline lib/bootloader_setup.pm]: Perhaps I wanted to catch the grub menu :) It safe to remove.

**Comment by okurz** [comment]
Any luck on updating? Just smaller changes for a start would be nice :-)

File: lib/registration.pm (hunk 2 above)
Comment (dgutu-suse) [inline lib/registration.pm]: I've found why I was using this next. Its safe to be removed.

File: products/sle/main.pm (hunk 5 above)
Comment (dgutu-suse) [inline products/sle/main.pm]: ok

File: tests/autoyast/autoyast_reboot.pm (hunk 6 above)
Comment (dgutu-suse) [inline tests/autoyast/autoyast_reboot.pm]: Actually this change is very related because autoupgrade jobs are also migration and I think they also deserves to be patched :) If autoupgrade are not migrations then I will revert changes to this file.

File: tests/update/patch_before_migration.pm (hunk 7 above)
Comment (dgutu-suse) [inline tests/update/patch_before_migration.pm]: ok

File: tests/update/patch_before_migration.pm (hunk 8 above)
Comment (dgutu-suse) [inline tests/update/patch_before_migration.pm]: +1

File: tests/x11/reboot_and_install.pm (hunk 13 above)
Comment (dgutu-suse) [inline tests/x11/reboot_and_install.pm]: it is evaluation of  lib/bootloader_setup   
###
if (get_var('ZDUP') || get_var("ONLINE_MIGRATION")) {                                                                                     send_key 'ret';    # boot from hard disk                                                                                              
return 3;
###

File: lib/utils.pm (hunk 4 above)
Comment (dgutu-suse) [inline lib/utils.pm]: Removing, changed mainly to handle sles11

File: lib/utils.pm (hunk 3 above)
Comment (dgutu-suse) [inline lib/utils.pm]: Removing, changed mainly to handle sles11

File: tests/update/patch_before_migration.pm (hunk 9 above)
Comment (dgutu-suse) [inline tests/update/patch_before_migration.pm]: Removing, was for sle11 to disable refreshing of repos after registering.

File: tests/update/patch_before_migration.pm (hunk 11 above)
Comment (dgutu-suse) [inline tests/update/patch_before_migration.pm]: +1

File: tests/update/patch_before_migration.pm (hunk 12 above)
Comment (dgutu-suse) [inline tests/update/patch_before_migration.pm]: ok

File: tests/update/patch_before_migration.pm (hunk 10 above)
Comment (dgutu-suse) [inline tests/update/patch_before_migration.pm]: ok

File: tests/x11/reboot_and_install.pm (hunk 14 above)
Comment (dgutu-suse) [inline tests/x11/reboot_and_install.pm]: Because of my (bad) qemu  setup, its safe to remove.

File: tests/x11/reboot_and_install.pm (hunk 15 above)
Comment (dgutu-suse) [inline tests/x11/reboot_and_install.pm]: ok

**Comment by okurz** [comment]
@dgutu-suse please update your PR with commits that focus on one aspect first so that we can merge them soon. Then we can continue with the next steps afterwards. I would like to see your changes in "master" so that we can make real use of it and others help maintaining it.

File: tests/boot/boot_to_desktop_sym.pm [hunk 16]
```diff
@@ -0,0 +1 @@
+boot_to_desktop.pm
```
Comment (mitiao) [inline tests/boot/boot_to_desktop_sym.pm]: why not directly call boot_to_desktop?

File: tests/console/consoletest_finish_sym.pm [hunk 17]
```diff
@@ -0,0 +1 @@
+consoletest_finish.pm
```
Comment (mitiao) [inline tests/console/consoletest_finish_sym.pm]: call consoletest_finish directly

File: tests/update/patch_before_migration.pm [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+    # stop packagekit service
+    script_run "systemctl mask packagekit.service";
+    script_run "systemctl stop packagekit.service";
+
+    assert_script_run("zypper lr && zypper mr --disable --all");
+    save_screenshot;
+    yast_scc_registration();
+    assert_script_run('zypper lr -d');
+    minimal_patch_system(version_variable => 'HDDVERSION');
```
Comment (mitiao) [inline tests/update/patch_before_migration.pm]: this only apply minimal patch, i don't see fully patch scenario

**Comment by mitiao** [comment]
Seems you have implemented upgrading on patched system by autoyast and zdup, what about upgrading by DVD?

File: tests/boot/boot_to_desktop_sym.pm (hunk 16 above)
Comment (dgutu-suse) [inline tests/boot/boot_to_desktop_sym.pm]: When a test is called 2 times the last is skipped.

File: tests/console/consoletest_finish_sym.pm (hunk 17 above)
Comment (dgutu-suse) [inline tests/console/consoletest_finish_sym.pm]: When a test is called 2 times the last is skipped.

File: tests/update/patch_before_migration.pm (hunk 18 above)
Comment (dgutu-suse) [inline tests/update/patch_before_migration.pm]: Thank you for taking a look :) I'm adding the variable to control the minimal and full patching.

**Comment by dgutu-suse** [comment]
@mitiao, can you please rephrase? autoupgrade by yast, zdup are all upgrade from DVD or do you mean another thing?

**Comment by mitiao** [comment]
testing like this: 
https://openqa.suse.de/tests/773488
but patched before upgrading.

**Comment by dgutu-suse** [comment]
@mitiao - yes, the one you show is not patching anything, here is the same example on my openqa instance: http://corr.arch.suse.de/tests/1514

File: lib/utils.pm [hunk 19]
```diff
@@ -343,9 +343,8 @@ sub reboot_gnome {
         assert_and_click 'reboot-auth-typed', 'right';
         assert_and_click 'reboot-auth-showtext';         # Click the 'Show Text' Option to enable the display of the typed text
         assert_screen 'reboot-auth-correct-password';    # Check the password is correct
-
-        # we need to kill ssh for iucvconn here,
-        # because after pressing return, the system is down
+                                                         # we need to kill ssh for iucvconn here,
+                                                         # because after pressing return, the system is down
```
Comment (okurz) [inline lib/utils.pm]: this comment indendation looks weird, please re-add the blank line before

File: products/sle/main.pm [hunk 20]
```diff
@@ -969,7 +969,12 @@ sub load_fips_tests_crypt() {
     loadtest "console/cryptsetup";
     loadtest "console/ecryptfs_fips";
 }
-
+sub load_patching_tests() {
```
Comment (okurz) [inline products/sle/main.pm]: please add blank line before

File: products/sle/main.pm [hunk 21]
```diff
@@ -969,7 +969,12 @@ sub load_fips_tests_crypt() {
...
+    loadtest 'boot/boot_to_desktop_sym';
+    loadtest 'update/patch_before_migration';
+    loadtest 'console/consoletest_finish_sym';
+    loadtest 'x11/reboot_and_install';
+}
 sub prepare_target() {
```
Comment (okurz) [inline products/sle/main.pm]: please add blank line in between

File: tests/console/consoletest_finish.pm [hunk 22]
```diff
@@ -24,9 +24,7 @@ sub run() {
     type_string "loginctl --no-pager\n";
     wait_still_screen(2);
     save_screenshot();
-
     script_run "systemctl unmask packagekit.service";
-
```
Comment (okurz) [inline tests/console/consoletest_finish.pm]: please keep blank lines as before

**Comment by mitiao** [comment]
with your given instance: http://corr.arch.suse.de/tests/1514, i think it is patched offline migration by zdup.
the scenario by DVD I expect is:
patch system -> reboot -> select 'Upgrade' in bootmenu -> perform upgrading by yast installer

**Comment by dgutu-suse** [comment]
@mitiao, my bad, I gave you the zdup case. Here is the migration from installer - http://corr.arch.suse.de/tests/1506

**Comment by QingmingSu** [comment]
Please don't make it a mandatory action to patch (update) the image before migrating. Because sometimes we'd like to test migration from an image with no patch (update).

**Comment by dgutu-suse** [comment]
@QingmingSu, the original hdd image will remain intact so you are able to re-use it for your purposes. Does my explanation make send to you?

File: tests/update/patch_before_migration.pm [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+use utils;
+use registration;
+
+sub patching_sle() {
+
+    select_console 'root-console';
+
+    set_var("VIDEOMODE",    'text');
+    set_var("SCC_REGISTER", 'installation');
+    set_var("SP2ORLATER",   '0') if (!sle_version_at_least('12-SP2', version_variable => 'HDDVERSION'));
```
Comment (dgutu-suse) [inline tests/update/patch_before_migration.pm]: @mitiao, I've encounter a problem to register scc addons in text mode after your changes here - https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/2138 
Without this I have troubles for sles12ga and sles12sp1.

File: tests/installation/zdup.pm [hunk 24]
```diff
@@ -90,6 +90,7 @@ sub run() {
                        [ -z \$dev ] || echo \"found dev \$dev with label \$label\"";
             # if that fails, e.g. if volume descriptor too long, just try /dev/sr0
             $defaultrepo = "dvd:/?devices=\${dev:-/dev/sr0}";
+            $defaultrepo = "cd:/?devices=\${dev:-/dev/sr0}" if get_var('PATCH');
```
Comment (dgutu-suse) [inline tests/installation/zdup.pm]: Without this change zypper_lr is complaining about missing repo and test fail. Could this be related to the fact that this testcase will use BOOTFROM=d variable?

File: tests/update/patch_before_migration.pm (hunk 23 above)
Comment (mitiao) [inline tests/update/patch_before_migration.pm]: version_variable => 'HDDVERSION' indicates that we are going to do registration on HDD images, so set_var should be HDD_SP2ORLATER.

**Comment by sysrich** [comment]
@QingmingSu the SLE PRD's make it quite clear that the expected supported behaviour is to support migrations with either a minimal patch (just the update stack) or a full patch.

I can find no mention in the PRD of supporting migrations without patching - if we don't support it, why do we need to test it?

**Comment by sysrich** [comment]
@dgutu-suse is this still WIP/notready?

File: lib/utils.pm [hunk 25]
```diff
@@ -395,6 +395,7 @@ sub reboot_gnome {
     assert_screen 'logoutdialog';
     assert_and_click 'logoutdialog-reboot-highlighted';
 
+
```
Comment (okurz) [inline lib/utils.pm]: unrelated change

File: tests/console/consoletest_finish.pm [hunk 26]
```diff
@@ -24,7 +24,6 @@ sub run() {
     type_string "loginctl --no-pager\n";
     wait_still_screen(2);
     save_screenshot();
-
```
Comment (okurz) [inline tests/console/consoletest_finish.pm]: unrelated change

File: tests/installation/zdup.pm [hunk 27]
```diff
@@ -65,7 +65,7 @@ sub run() {
     }
     else {
         #SUSEMIRROR not set, zdup from ftp source for online migration
-        if (check_var('TEST', "migration_zdup_online_sle12_ga")) {
+        if (get_var('TEST') =~ /migration_zdup_online_sle12_ga/) {
```
Comment (okurz) [inline tests/installation/zdup.pm]: why not keep the old?

File: tests/update/patch_before_migration.pm [hunk 28]
```diff
@@ -0,0 +1,76 @@
...
+# Maintainer: Dumitru Gutu <dgutu@suse.de>
+
+use base "consoletest";
+use strict;
+use testapi;
+use utils;
+use registration;
+
+sub patching_sle() {
+
```
Comment (okurz) [inline tests/update/patch_before_migration.pm]: delete blank line

File: tests/update/patch_before_migration.pm [hunk 29]
```diff
@@ -0,0 +1,76 @@
...
+        die "System is still registered" unless $output =~ /Not Registered/;
+        save_screenshot;
+    }
+    assert_script_run("zypper mr --enable --all");
+    set_var("VIDEOMODE", '');
+    if (get_var('TEST') !~ /migration_offline_sle12sp1_smt/) { set_var("SCC_REGISTER", ''); }
+}
+
+sub run() {
+
```
Comment (okurz) [inline tests/update/patch_before_migration.pm]: delete blank line

File: tests/update/patch_before_migration.pm [hunk 30]
```diff
@@ -0,0 +1,76 @@
...
+    select_console 'root-console';
+
+    type_string "chown $username /dev/$serialdev\n";
```
Comment (okurz) [inline tests/update/patch_before_migration.pm]: this is repeating what is also necessary in consoletest_setup. Please extract method

**Comment by QingmingSu** [comment]
@sysrich, Agree that it makes more sense to always update the system before migrating to newer version.

File: lib/utils.pm (hunk 25 above)
Comment (dgutu-suse) [inline lib/utils.pm]: Removing...

File: tests/console/consoletest_finish.pm (hunk 26 above)
Comment (dgutu-suse) [inline tests/console/consoletest_finish.pm]: Adding back then.

File: tests/installation/zdup.pm (hunk 27 above)
Comment (dgutu-suse) [inline tests/installation/zdup.pm]: Because every migration testcase will have fullupdate and minimalupdate in their names, or shorten one like mu (minimalupdate) and fu (fullupdate)

File: tests/installation/zdup.pm (hunk 24 above)
Comment (dgutu-suse) [inline tests/installation/zdup.pm]: I'm gonna remove this "if" because the issue is in my openqa environment.

File: tests/update/patch_before_migration.pm (hunk 28 above)
Comment (dgutu-suse) [inline tests/update/patch_before_migration.pm]: OK.

File: tests/update/patch_before_migration.pm (hunk 23 above)
Comment (dgutu-suse) [inline tests/update/patch_before_migration.pm]: It's even save to be removed. Thx for noticing it.

File: tests/update/patch_before_migration.pm (hunk 29 above)
Comment (dgutu-suse) [inline tests/update/patch_before_migration.pm]: Am I fan of blank lines ? :) Removing.

File: tests/update/patch_before_migration.pm (hunk 30 above)
Comment (dgutu-suse) [inline tests/update/patch_before_migration.pm]: ok

**Comment by dgutu-suse** [comment]
@okurz @sysrich @mitiao @QingmingSu thx for reviewing. I have only one things that stops me to remove the labels. On my openqa instance the mounted installation source (like sle12sp3 dvd) is always a CD-ROM but as soon as I see on o.s.d is always DVD-ROM so my runs failing on zypper_lr because of mismatching or repos. So I've removed this part - $defaultrepo = "cd:/?devices=\${dev:-/dev/sr0}" if get_var('PATCH'); I get also DVD-ROM when booting the same hdd image through virt-manager so there is something that makes it CD-ROM in my openqa instance. 
Some examples:
http://corr.arch.suse.de/tests/1746#step/zypper_lr/13 - here the source was still mounted as CD-ROM that's why it passing the zdup test but since I removed the if above it will start to fail on zdup testcase.

**Comment by dgutu-suse** [comment]
All above applying only to 'zdup'. autoupgrade and simple offline migration is installed from CD-ROM.

**Comment by sysrich** [comment]
@dgutu-suse what is the question? - your requestion from 13 days ago doesn't have a question mark in it (besides the the defaultrepo string) - so I can't identify what you want answered.

**Comment by sysrich** [comment]
on hold till @dgutu-suse is back from FTO, has a modification he wishes to make

**Comment by dgutu-suse** [comment]
It's ok to be merged.
